File: gui_apps/PyDoodle/step06/drawWindow.py
"""
// ============================================================================
// drawWindow.py: custom QMainWindow derived class for main window
//
// Tutorial - PyQt5 Doodle Application
// Based on a similar tutorial for Borland ObjectWindows Library (OWL) 
// @author: Manish Bhobe
// My experiments with the Qt Framework. Use at your own risk!!
// ============================================================================
"""
import sys
import os
import pathlib
from PyQt5.QtCore import *
from PyQt5.QtGui import *
from PyQt5.QtWidgets import *
from squiggle import *
import darkdetect

sys.path.append(os.path.join(pathlib.Path(__file__).parents[1], 'common'))
import mypyqt5_utils as utils


class DrawWindow(QMainWindow):
    def __init__(self, *args, **kwargs):
        super(QMainWindow, self).__init__(*args, **kwargs)
        self.setWindowTitle("PyQt5 Doodle - Step06: Drawing multiple Squiggles with own pen width & color")
        # self.setStyleSheet("background-color: white")
        self.setGeometry(QRect(100, 100, 640, 480))
        self.modified = False
        self.squiggles = []
        self.dragging = False
        self.penColor = QColor(qRgb(0, 65, 255))
        self.penWidth = 3
        self.currSquiggle = None
        self.setMouseTracking(True)

    def drawSquiggles(self, painter: QPainter) -> None:
        for squiggle in self.squiggles:
            squiggle.draw(painter)

    # operating system Events
    def closeEvent(self, e):
        """ called just before the main window closes """
        if self.modified:
            resp = QMessageBox.question(self, "Confirm Close",
                                        "This will close the application.\nOk to quit?",
                                        QMessageBox.Yes | QMessageBox.No, QMessageBox.No)
            if resp == QMessageBox.Yes:
                e.accept()
            else:
                e.ignore()
        else:
            e.accept()

    def paintEvent(self, e: QPaintEvent) -> None:
        """ handler for paint events """
        print('In DrawWindow::paintEvent() - ', end='', flush=True)
        if darkdetect.isDark():
            print('dark theme detected - ', end='', flush=True)
            utils.setDarkPalette(QApplication.instance())
        else:
            print('light theme detected - ', end='', flush=True)
            # QApplication.instance().setStyleSheet("")
            utils.setDarkPalette(QApplication.instance(), False)
        painter = QPainter()
        try:
            width, height = self.width(), self.height()
            color = QColor(53, 53, 53) if darkdetect.isDark() else utils.getPaletteColor(QPalette.Window)
            print(f'QPalette.Window color = {color.name()}')
            painter.begin(self)
            painter.setBrush(color)
            painter.setPen(color)
            painter.drawRect(0, 0, width, height)
            painter.setRenderHint(QPainter.Antialiasing, True)
            self.drawSquiggles(painter)
        finally:
            painter.end()

    def mousePressEvent(self, e: QMouseEvent) -> None:
        """ handler for mouse press (left or right button) events """
        if e.button() == Qt.LeftButton:
            if (e.modifiers() & Qt.ControlModifier):
                # if Ctrl key is also pressed with mouse press, display
                # dialog to change pen thickness
                newWidth, ok = QInputDialog.getInt(self, "Pen Width",
                                                   "Enter new pen width (2-12):",
                                                   self.penWidth, 2, 12)
                if ok:  # user clicked Ok on QInputDialog
                    self.penWidth = newWidth
            else:
                # start a new line
                #self.currSquiggle = Squiggle(self.penWidth, self.penColor)
                self.currSquiggle = Squiggle()
                self.currSquiggle.penWidth = self.penWidth
                self.currSquiggle.penColor = self.penColor
                self.squiggles.append(self.currSquiggle)
                pt = QPoint(e.pos().x(), e.pos().y())
                self.currSquiggle.append(pt)
                self.modified = True
                self.dragging = True
        elif e.button() == Qt.RightButton:
            if (e.modifiers() & Qt.ControlModifier):
                # if Ctrl key is also pressed with mouse press, display
                # dialog to change pen color
                newColor = QColorDialog.getColor(self.penColor, self)
                if newColor.isValid():
                    self.penColor = newColor
            else:
                for squiggle in self.squiggles:
                    squiggle = None
                self.squiggles = []
                self.modified = False
                self.update()

    def mouseMoveEvent(self, e: QMouseEvent) -> None:
        """ handler for mouse drag (left or right button) events
            NOTE: you must call setMouseTracking(True) so window can receive mouse drag events
        """
        if (e.buttons() == Qt.LeftButton) and (self.dragging):
            assert self.currSquiggle != None, "FATAL: self.currLine is None, when expecting valid!"
            pt = QPoint(e.pos().x(), e.pos().y())
            self.currSquiggle.append(pt)
            self.update()
        else:
            e.accept()

    def mouseReleaseEvent(self, e: QMouseEvent) -> None:
        """ handler for mouse (left or right button) released events """
        if (e.button() == Qt.LeftButton) and (self.dragging):
            assert self.currSquiggle != None, "FATAL: self.currLine is None, when expecting valid!"
            pt = QPoint(e.pos().x(), e.pos().y())
            self.currSquiggle.append(pt)
            self.dragging = False
            self.update()
        else:
            e.accept()

File: gui_apps/Doodle/step05/DrawWindow.cc
// DrawWindow.cc: implements DrawWindow class
#include "common.hxx"
#include "DrawWindow.h"
#include "Line.h"
#include <QApplication>
#include <QColorDialog>
#include <QInputDialog>
#include <QMessageBox>
#include <QtGui>

const QString AppTitle("Qt Scribble");
const QString WindowTitle("Qt Doodle - Step05: Changing Line color & thickness");

DrawWindow::DrawWindow()
{
   setAttribute(Qt::WA_StaticContents);
   setWindowTitle(WindowTitle);
   resize(640, 480);
   _modified = false;
   _dragging = false;
   _penWidth = 3;
   _penColor = qRgb(0, 0, 255);
   _line = nullptr;
}

DrawWindow::~DrawWindow() { delete _line; }

void DrawWindow::closeEvent(QCloseEvent *event)
{
   // window is about to close, prompt user & decide if ok to quit
   // based on user's response.
   if (_modified) {
      switch (QMessageBox::question(this, tr("Qt Scribble Tutorial"),
                                    tr("The doodle has been modified.\nDo you want to close without saving?"),
                                    QMessageBox::Yes | QMessageBox::No, QMessageBox::No)) {
         case QMessageBox::Yes:
            // ok to quit
            qDebug() << "User chose to quit without saving!";
            event->accept();
            break;
         default:
            // don't quit yet
            event->ignore();
      }
   }
}

void DrawWindow::drawLineTo(const QPoint &pt)
{
   // draw line from _lastPt to pt
   QPainter painter(&_image);
   QPen pen(_penColor, _penWidth);
   painter.setRenderHint(QPainter::Antialiasing);
   painter.setPen(pen);
   painter.drawLine(_lastPt, pt);
   _lastPt = pt;
   update();
}

void DrawWindow::clearImage()
{
   QColor color = getPaletteColor(QPalette::Window);
   _image.fill(color);
   //_image.fill(qRgb(255, 255, 255));
   update();
}

void DrawWindow::mousePressEvent(QMouseEvent *event)
{
   if (event->button() == Qt::LeftButton) {
      // left mouse button pressed
      if (QApplication::keyboardModifiers() & Qt::ControlModifier) {
         // display message box & get new width of pen
         // NOTE: this is not the best way to call up a dialog!!! It'll do for now
         bool ok;
         int newPenWidth = QInputDialog::getInt(this, AppTitle, QString("Enter new pen width (2-12):"),
                                                _line->penWidth(), 2, 12, 1, &ok);
         if (ok) {
            qDebug() << "New pen width selected: " << newPenWidth;
            _penWidth = newPenWidth;
         }
      } else {
         delete _line;
         _line = nullptr;
         clearImage();
         _lastPt = event->pos();
         _dragging = true;
         _modified = true;
         _line = new Line(_penWidth, _penColor);
         _line->addPoint(event->pos());
      }
   } else if (event->button() == Qt::RightButton) {
      if (QApplication::keyboardModifiers() & Qt::ControlModifier) {
         // if Ctrl key is also pressed, display standard color
         // dialog & get new pen color
         QColor color = QColorDialog::getColor(_penColor, this);
         if (color.isValid())
            _penColor = color;

      } else {
         clearImage();
         _modified = false;
      }
   }
}

void DrawWindow::mouseMoveEvent(QMouseEvent *event)
{
   if (event->buttons() & Qt::LeftButton && _dragging) {
      drawLineTo(event->pos());
      _line->addPoint(event->pos());
   }
}

void DrawWindow::mouseReleaseEvent(QMouseEvent *event)
{
   if ((event->button() == Qt::LeftButton) && _dragging) {
      drawLineTo(event->pos());
      _line->addPoint(event->pos());
      _dragging = false;
   }
}

void DrawWindow::resizeEvent(QResizeEvent *event)
{
   if (width() > _image.width() || height() > _image.height()) {
      int newWidth = qMax(width(), _image.width());
      int newHeight = qMax(height(), _image.height());
      resizeImage(QSize(newWidth, newHeight));
      update();
   }
   QWidget::resizeEvent(event);
}

void DrawWindow::paintEvent(QPaintEvent *event)
{
   QPainter painter(this);
   painter.setRenderHint(QPainter::Antialiasing);
   QRect dirtyRect = event->rect();
   painter.drawImage(dirtyRect, _image, dirtyRect);
}

void DrawWindow::resizeImage(const QSize &newSize)
{
   if (_image.size() == newSize)
      return;
   QImage newImage(newSize, QImage::Format_RGB32);
   QColor color = getPaletteColor(QPalette::Window);
   newImage.fill(color);
   // newImage.fill(qRgb(255, 255, 255));

   // draw existing image over new image
   QPainter painter(&newImage);
   painter.setRenderHint(QPainter::Antialiasing);
   painter.drawImage(QPoint(0, 0), _image);
   _image = newImage;
}
